### PR TITLE
feat: Implement tool for parallel execution

### LINKED ISSUE #2001: Implement tool for parallel execution
Implement a library module that is meeting the requirements described in epic.

### PR DESCRIPTION
**Pull request by jan-gogo**
Fixes #2001

### Parallel execution algorithm

* [README.md](https://github.com/Flank/flank/blob/2001_Implement_tool_for_parallel_execution/tool/execution/parallel/README.md)

#### Example graph execution

example-graph

#### Algorithm

parallel-execution

#### Functions

parallel-execution

#### Types

parallel-execution

### Different graph states

parallel-execution

Unit tests are passing.

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: tool/execution/parallel/README.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    * Execution - [Execute.kt](./src/main/kotlin/flank/exection/parallel/Execute.kt)
+    * Task factories DSL - [Create.kt](./src/main/kotlin/flank/exection/parallel/Create.kt)
+    * Graph validation - [Create.kt](./src/main/kotlin/flank/exection/parallel/Validate.kt)
+    * Graph reduction - [Reduce.kt](./src/main/kotlin/flank/exection/parallel/Reduce.kt)
+* Example usage - [Example.kt](./src/test/kotlin/flank/exection/parallel/Example.kt)
+
+# Design
+
+Imagine a complicated long-running `execution` that needs to perform several operations (`tasks`) in correct order to
+collect a required `data` and produce `side effects`. Any `execution` like that, can be modeled as set of unrelated
```
Comment (Sloox) [inline tool/execution/parallel/README.md:20]:
```suggestion
collect required `data` and produce `side effects`. Any `execution` like that, can be modeled as a set of unrelated
```
data is plural in this case and doesnt need `a` :) And set in this case needs it :'(

File: tool/execution/parallel/README.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+object Baz : Parallel.Type<Boolean>
+object Logger : Parallel.Type<Output>
+```
+
+## Task
+
+* is representing a single atomic operation in the execution process.
+* is a relation of signature and task function.
+* `arguments` for task are coming from `initial state` (`arguments`) or results from other `tasks`.
+* can be uniquely identified using return type of its signature.
```
Comment (Sloox) [inline tool/execution/parallel/README.md:51]:
```suggestion
* can be uniquely identified using the return type of its signature.
```

File: tool/execution/parallel/README.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+### Signature
+
+* is a set of argument `types` related with one return `type` just like in normal functions.
```
Comment (Sloox) [inline tool/execution/parallel/README.md:55]:
```suggestion
* is a set of argument `types` related with one return `type` just like in a normal functions.
```

**Review by Sloox — COMMENTED** [review]
Part one of simple review as its a large review to do.

**Review by pawelpasterz — APPROVED** [review]

**Review by Sloox — APPROVED** [review]
This is a very large PR. Phew :worried:

**Pull request merged**